### DISCUSSION TITLE
Action fails on 'application_terminated' event

### DISCUSSION THREAD

**Discussion by noib3** [asmvik/yabai#620] [Q&A]
I'm using the following code to automatically destroy an empty space

````
read -r -d '' destroy_space_if_empty <<- 'EOF'
active_space_index=$(yabai -m query --spaces --space | jq .index)
if $(yabai -m query --windows --space ${active_space_index} | jq 'length == 0'); then
    yabai -m space ${active_space_index} --destroy
fi
EOF

yabai -m signal --add event=window_destroyed action="${destroy_space_if_empty}"
yabai -m signal --add event=application_terminated action="${destroy_space_if_empty}"
````
the action associated to the `window_destroyed` event works without problems, but the one associated to the `application_terminated` event doesn't seem to do anything.

I've tested different actions with the same event and they all work, for example
````
read -r -d '' destroy_space_if_empty <<- 'EOF'
active_space_index=$(yabai -m query --spaces --space | jq .index)
echo ${active_space_index} >> ~/spaces
EOF

yabai -m signal --add event=application_terminated action="${destroy_space_if_empty}"
````
creates the file `spaces` in the home directory and `cat` returns the correct space.

Any ideas?

**Comment by dominiklohmann** [comment]
`window_destroyed` always triggers before `application_terminated` does for applications that have own a window and are thus relevant to yabai.

You can verify this easily using these two signals:

```sh
yabai -m signal --add event=window_destroyed action="osascript -e 'display notification \"window_destroyed\"'"
yabai -m signal --add event=application_terminated action="osascript -e 'display notification \"application_terminated\"'"
```

**Comment by noib3** [comment]
@dominiklohmann if I quit a program with `cmd-q` instead of closing one of its windows (be it the last one or not, doesn't matter) with `cmd-w`, only `application_terminated` triggers, `window_destroyed` doesn't.

In theory this shouldn't be a problem: if I link the action `destroy_space_if_empty` to both `application_terminated` and `window_destroyed` I should cover both cases:

a) I close the last window of a space without quitting the program -> `window_destroyed` is triggered;

b) I quit a program and there are no more windows left on that space -> `application_terminated` is triggered.

In practice, it seems that the space is only destroyed if I close the last window without quitting the program.
